posthog/year_in_posthog/calculate_2022.py

```diff
@@ -0,0 +1,178 @@
+                count(*) AS viewed_recording_count
+            FROM
+                posthog_sessionrecordingviewed
+            WHERE
+                date_part('year', created_at) = 2022
+                AND user_id = %s
+            GROUP BY
+                user_id
+        ) AS recording_stats_inner
+),
+experiments_stats AS (
+    SELECT
+        user_id,
+        experiments_created_count,
+        CASE
+            WHEN experiments_created_count >= 4 THEN 'scientist'
+        END AS badge
+    FROM
+        (
+            SELECT
+                created_by_id AS user_id,
+                count(*) AS experiments_created_count
+            FROM
+                posthog_experiment
+            WHERE
+                date_part('year', created_at) = 2022
+                AND created_by_id = %s
+            GROUP BY
+                user_id
+        ) AS experiment_stats_inner
+),
+dashboards_created_stats AS (
+    SELECT
+        user_id,
+        dashboards_created_count,
+        CASE
+            WHEN dashboards_created_count >= 10 THEN 'curator'
+        END AS badge
+    FROM
+        (
+            SELECT
+                created_by_id AS user_id,
+                count(*) AS dashboards_created_count
+            FROM
+                posthog_dashboard
+            WHERE
+                date_part('year', created_at) = 2022
+                AND created_by_id = %s
+            GROUP BY
+                user_id
+        ) AS dashboard_stats_inner
+)
+SELECT
+    id,
+    array_remove(ARRAY [
+        case when posthog_user.last_login >= '1-Jan-2022' then 'astronaut' end,
+        insight_stats.badge,
+        flag_stats.badge,
+        recording_viewed_stats.badge,
+        experiments_stats.badge,
+        dashboards_created_stats.badge,
+        case when
+            recording_viewed_stats.badge is not null
+                and flag_stats.badge is not null
+                and insight_stats.badge is not null
+            then 'champion' end
+    ], NULL) AS badges,
+    insight_stats.insight_created_count,
+    flag_stats.flag_created_count,
+    recording_viewed_stats.viewed_recording_count,
+    experiments_stats.experiments_created_count,
+    dashboards_created_stats.dashboards_created_count
+FROM
+    posthog_user
+    LEFT JOIN insight_stats ON posthog_user.id = insight_stats.user_id
+    LEFT JOIN flag_stats ON posthog_user.id = flag_stats.user_id
+    LEFT JOIN recording_viewed_stats ON posthog_user.id = recording_viewed_stats.user_id
+    LEFT JOIN experiments_stats ON posthog_user.id = experiments_stats.user_id
+    LEFT JOIN dashboards_created_stats ON posthog_user.id = dashboards_created_stats.user_id
+WHERE posthog_user.id = %s
+"""
+
+
+def dictfetchall(cursor):
+    "Return all rows from a cursor as a dict"
+    columns = [col[0] for col in cursor.description]
+    return [dict(zip(columns, row)) for row in cursor.fetchall()]
+
+
+@timed("year_in_posthog_2022_calculation")
+@cache_for(timedelta(seconds=30))
```

Comment:
Why cache this so un-aggressively?

This caching applies only to a single server (in-memory) but this computation seems quite harsh.

Reply:
Partly this feels like premature optimization because the query is really fast. Just don't want a public endpoint to be able to spam the database.

Partly so that if we release it and discover bugs I can still roll out changes without waiting for a painful TTL (and then turn up the cache if necessary)